backend/kitschy_api/tests/product_tests.py

```diff
@@ -0,0 +1,184 @@
+            "desc": "Pre-Order Description",
+            "price": 30.00,
+            "stock": -10,
+        }
+        response = authenticated_user.post("/api/products/", product)
+
+        assert response.status_code == 201
+        assert response.data["name"] == product["name"]
+
+    def test_update_pre_order_product(self, authenticated_user):
+        product = baker.make("Product", stock=-10)
+        response = authenticated_user.put(
+            f"/api/products/{product.product_id}/",
+            {
+                "name": "Updated Pre-Order Product",
+                "desc": "Updated Pre-Order Description",
+                "price": 40.00,
+                "stock": -20,
+            },
+        )
+
+        assert response.status_code == 200
+
+    def test_delete_pre_order_product(self, authenticated_user):
+        product = baker.make("Product", stock=-10)
+        response = authenticated_user.delete(
+            f"/api/products/{product.product_id}/"
+        )
+
+        assert response.status_code == 204
+        assert not response.data
+
+
+# Tests for invalid data
+@pytest.mark.django_db
+class TestInvalidProductAPI:
+
+    endpoint = "/api/products/"
+
+    def test_create_invalid_product(self, authenticated_user):
+        product = {
+            "name": "Invalid Product",
+            "desc": "Invalid Description",
+            "price": -10.00,
+            "stock": 10,
+        }
+        response = authenticated_user.post("/api/products/", product)
+
+        assert response.status_code == 400
+
+    def test_update_invalid_product(self, authenticated_user):
+        product = baker.make("Product")
+        response = authenticated_user.put(
+            f"/api/products/{product.product_id}/",
+            {
+                "name": "Updated Invalid Product",
+                "desc": "Updated Invalid Description",
+                "price": -20.00,
+                "stock": 20,
+            },
+        )
+
+        assert response.status_code == 400
+
+    def test_delete_invalid_product(self, authenticated_user):
+        invalid_product = baker.make("Product", price=-10)
+        response = authenticated_user.delete(
+            f"/api/products/{invalid_product.product_id}/"
+        )
+
+        assert response.status_code == 204
+        assert not response.data
+
+
+# Tests for edge cases
+@pytest.mark.django_db
+class TestProductEdgeCases:
+
+    endpoint = "/api/products/"
+
+    def test_create_empty_product(self, authenticated_user):
+        product = {}
+        response = authenticated_user.post("/api/products/", product)
+
+        assert response.status_code == 400
+
+    def test_update_empty_product(self, authenticated_user):
+        product = baker.make("Product")
+        response = authenticated_user.put(
+            f"/api/products/{product.product_id}/",
+            {},
+        )
+
+        assert response.status_code == 400
+
+    def test_delete_empty_product(self, authenticated_user):
```

Comment:
Redundant test. This is the same as line 63, as it is impossible to have an empty `Product` in the database.